pkg/common/cns-lib/volume/manager.go

```diff
@@ -1230,18 +1274,48 @@ func (m *defaultManager) expandVolumeWithImprovedIdempotency(ctx context.Context
 		}
 	}
 
-	// Get the taskInfo.
-	taskInfo, err := cns.GetTaskInfo(ctx, task)
-	if err != nil || taskInfo == nil {
-		log.Errorf("failed to get taskInfo for ExtendVolume task from vCenter %q with err: %v",
-			m.virtualCenter.Config.Host, err)
-		if err != nil {
-			faultType = ExtractFaultTypeFromErr(ctx, err)
-		} else {
-			faultType = csifault.CSITaskInfoEmptyFault
+	taskInfo, err := task.WaitForResult(ctx, nil)
+	if err != nil {
+		log.Debugf("Failed to get ExtendVolume task result for task %s with error %v. "+
+			"Querying CNS to determine if the volume %s was successfully created.",
+			task.Reference().Value, err, volumeID)
+		// WaitForResult can fail for many reasons, including:
+		// - Task was completed and deleted from VCDB.
+		// - CNS restarted and marked "InProgress" tasks as "Failed".
+		// - Any other CNS failures.
+		// - Any other failures at lower layers like FCD and SPS.
+		// In all cases, query CNS to determine if the volume was expanded successfully.
+		queryFilter := cnstypes.CnsQueryFilter{
+			VolumeIds: []cnstypes.CnsVolumeId{{Id: volumeID}},
+		}
+		querySelection := cnstypes.CnsQuerySelection{
+			Names: []string{
+				string(cnstypes.QuerySelectionNameTypeBackingObjectDetails),
+			},
 		}
-		return faultType, err
+		queryResult, queryAllVolumeErr := m.QueryAllVolume(ctx, queryFilter, querySelection)
+		if queryAllVolumeErr != nil {
+			log.Debugf("failed to query CNS for volume %s with error: %v. Cannot "+
+				"determine if ExtendVolume task %s was successful.", volumeID,
+				queryAllVolumeErr, task.Reference().Value)
+			return ExtractFaultTypeFromErr(ctx, err), err
+		}
+		if len(queryResult.Volumes) > 0 &&
```

Comment:
What if we get multiple volumes? Can we log a separate log message to indicate the problem? It is rare, but its worth handling this as our stack does not guarantee unique volume names.

Reply:
changed the logic. If multiple entries are returned, we log that there are multiple entries and then check the capacity of the first one